src/pages/Dashboard/MintRepayVai/RepayVai/index.tsx

```diff
@@ -0,0 +1,142 @@
+import { AuthContext } from 'context/AuthContext';
+import { SecondaryButton, LabeledInlineContent, TokenTextField } from 'components';
+import { useVaiUser } from 'hooks/useVaiUser';
+import { useRepayVai } from 'clients/api';
+import toast from 'components/Basic/Toast';
+import useConvertToReadableCoinString from '../useConvertToReadableCoinString';
+import { VAI_SYMBOL } from '../constants';
+import { useStyles } from '../styles';
+
+export interface IRepayVaiUiProps {
+  disabled: boolean;
+  isRepayVaiLoading: boolean;
+  onSubmit: (value: BigNumber) => void;
+  userWeiBalance?: BigNumber;
+  mintedWei?: BigNumber;
+}
+
+export const RepayVaiUi: React.FC<IRepayVaiUiProps> = ({
+  disabled,
+  userWeiBalance,
+  mintedWei,
+  isRepayVaiLoading,
+  onSubmit,
+}) => {
+  const limitWei = React.useMemo(
+    () =>
+      userWeiBalance && mintedWei ? BigNumber.minimum(userWeiBalance, mintedWei) : new BigNumber(0),
+    [userWeiBalance?.toString(), mintedWei?.toString()],
+  );
+
+  const styles = useStyles();
+
+  // Convert minted wei into VAI
+  const readableRepayableVai = useConvertToReadableCoinString({
+    valueWei: mintedWei,
+    tokenSymbol: VAI_SYMBOL,
+  });
+
+  const hasRepayableVai = mintedWei?.isGreaterThan(0) || false;
+
+  return (
+    <AmountForm onSubmit={onSubmit} css={styles.tabContentContainer}>
+      {({ values, setFieldValue, handleBlur, isValid, dirty }) => (
+        <>
+          <div css={styles.ctaContainer}>
+            <TokenTextField
+              name="amount"
+              css={styles.textField}
+              tokenSymbol={VAI_SYMBOL}
+              value={values.amount}
+              onChange={amount => setFieldValue('amount', amount, true)}
+              onBlur={handleBlur}
+              maxWei={limitWei}
+              disabled={disabled || isRepayVaiLoading || !hasRepayableVai}
+              rightMaxButtonLabel="MAX"
+            />
+
+            <LabeledInlineContent
+              css={styles.getRow({ isLast: true })}
+              iconName={VAI_SYMBOL}
+              label="Repay VAI balance"
+            >
+              {readableRepayableVai}
+            </LabeledInlineContent>
+          </div>
+
+          <SecondaryButton
+            type="submit"
+            loading={isRepayVaiLoading}
+            disabled={disabled || !isValid || !dirty}
+            fullWidth
+          >
+            Repay VAI
+          </SecondaryButton>
+        </>
+      )}
+    </AmountForm>
+  );
+};
+
+const RepayVai: React.FC = () => {
+  const { account } = React.useContext(AuthContext);
+
+  const { userVaiMinted, userVaiBalance } = useVaiUser();
+
+  const { mutate: repayVai, isLoading: isRepayVaiLoading } = useRepayVai({
+    onError: error => {
+      toast.error({ title: error.message });
+    },
+    onSuccess: (_data, variables) => {
+      // @TODO: display success modal instead of toast once it's been
+      // implemented
+      toast.success({
```

Comment:
A strategy I've been following is to keep the container pure of all UI interactions. This is why I introduced the toastError. So the container could throw the error and we catch it and handle the appropriate UI updates.

I think this should also make testing nicer because we can test for errors on the containers and not UI interactions.

Reply:
Ok I see, that's a nice pattern.

Since our fetch functions already throw errors, I can use `mutateAsync` inside the UI component here to catch errors and display a toast (on success too).

Reply:
By the way, this PR originally branched out of another PR I was working and that I needed the work of. Now that's it's been merged in and this PR is rebased with `develop`, these changes no longer appear.
I've made the changes related to your comment in that PR: https://github.com/VenusProtocol/venus-protocol-interface/pull/282